File: apkEditor/src/main/java/com/mcal/apkeditor/AppListAdapter.java
package com.mcal.apkeditor;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.util.LruCache;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

public class AppListAdapter extends BaseAdapter {

    private final Context ctx;
    private final PackageManager pm;
    private final LruCache<String, Drawable> pkgDrawables = new LruCache<String, Drawable>(
            32) {
        protected void entryRemoved(boolean evicted, String key,
                                    @NonNull Drawable oldValue, Drawable newValue) {
            oldValue.setCallback(null);
        }
    };
    public List<AppInfo> appList = new ArrayList<AppInfo>();
    private APPLIST_ORDER order;

    public AppListAdapter(@NonNull Context ctx) {
        this.ctx = ctx;
        this.pm = ctx.getPackageManager();
    }

    public void setAppList(List<AppInfo> appList, @NonNull String order) {
        String[] orderConsts = ctx.getResources().getStringArray(
                R.array.order_value);
        if (order.equals(orderConsts[0])) {
            this.order = APPLIST_ORDER.BY_NAME;
        } else if (order.equals(orderConsts[1])) {
            this.order = APPLIST_ORDER.BY_INSTALL_TIME;
        } else {
            this.order = APPLIST_ORDER.BY_NAME;
        }
        sortAppList(appList);
        synchronized (this.appList) {
            this.appList.clear();
            this.appList.addAll(appList);
        }
    }

    private void sortAppList(List<AppInfo> appList) {
        final Locale locale = Locale.getDefault();
        Comparator<AppInfo> comparator = null;

        switch (order) {
            case BY_NAME:
                comparator = Comparator.comparing(arg0 -> arg0.appName.toLowerCase(locale));
                break;
            case BY_INSTALL_TIME:
                comparator = (arg0, arg1) -> arg0.lastUpdateTime < arg1.lastUpdateTime ? 1 : -1;
                break;
        }

        appList.sort(comparator);
    }

    public List<AppInfo> getAppList() {
        synchronized (this.appList) {
            List<AppInfo> retList = new ArrayList<>(appList);
            return retList;
        }
    }

    @Override
    public int getCount() {
        synchronized (this.appList) {
            return this.appList.size();
        }
    }

    @Override
    public Object getItem(int arg0) {
        synchronized (this.appList) {
            return this.appList.get(arg0);
        }
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @SuppressLint("InflateParams")
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final AppInfo appInfo = (AppInfo) getItem(position);
        if (appInfo == null) {
            return null;
        }

        ViewHolder viewHolder;
        if (convertView == null) {
            convertView = LayoutInflater.from(ctx).inflate(R.layout.item_applist, null);

            viewHolder = new ViewHolder();
            viewHolder.icon = convertView
                    .findViewById(R.id.app_icon);
            viewHolder.appName = convertView
                    .findViewById(R.id.app_name);
            viewHolder.desc1 = convertView
                    .findViewById(R.id.app_desc1);
            viewHolder.desc2 = convertView
                    .findViewById(R.id.app_desc2);

            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        try {
            viewHolder.appName.setText(appInfo.appName);

            // if (appInfo.isSysApp)
            // viewHolder.appName.setTextColor(0xfff75343);
            // else
            // viewHolder.appName.setTextColor(0xff0028c6);

            viewHolder.desc1.setText(appInfo.packagePath);

            Drawable icon = pkgDrawables.get(appInfo.packagePath);
            if (icon == null) {
                icon = appInfo.applicationInfo.loadIcon(pm);
                pkgDrawables.put(appInfo.packagePath, icon);
            }

            viewHolder.icon.setImageDrawable(icon);
        } catch (Throwable t) {
            t.printStackTrace();
        }

        // appCustomize.setupLongClickListener(convertView, appInfo);

        // appCustomize.setupClickListener(convertView, appInfo);

        return convertView;
    }

    private enum APPLIST_ORDER {
        BY_NAME, BY_INSTALL_TIME
    }

    static class ViewHolder {
        public ImageView icon;
        public TextView desc2;
        public TextView desc1;
        public TextView appName;
    }
}

File: apkEditor/src/main/java/com/mcal/apkeditor/AppListAdapter.kt
package com.mcal.apkeditor

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import java.util.*

class AppListAdapter(private val ctx: Context) : BaseAdapter() {
    private val pm: PackageManager = ctx.packageManager

    var appList = mutableListOf<AppInfo>()
    private var order: AppListOrder? = null
    fun setAppList(appList: MutableList<AppInfo>, order: String) {
        val orderConst = ctx.resources.getStringArray(
            R.array.order_value
        )
        when (order) {
            orderConst[0] -> {
                this.order = AppListOrder.BY_NAME
            }
            orderConst[1] -> {
                this.order = AppListOrder.BY_INSTALL_TIME
            }
            else -> {
                this.order = AppListOrder.BY_NAME
            }
        }

        sortAppList(appList)
        synchronized(this.appList) {
            this.appList.clear()
            this.appList.addAll(appList)
        }
    }

    private fun sortAppList(appList: MutableList<AppInfo>) {
        val locale = Locale.getDefault()
        var comparator: Comparator<AppInfo>
        order?.let { appListOrder ->
            comparator = when (appListOrder) {
                AppListOrder.BY_NAME -> {
                    Comparator.comparing { arg0: AppInfo -> arg0.appName.lowercase(locale) }
                }
                AppListOrder.BY_INSTALL_TIME -> {
                    Comparator { arg0: AppInfo, arg1: AppInfo -> if (arg0.lastUpdateTime < arg1.lastUpdateTime) 1 else -1 }
                }
            }
            appList.sortWith(comparator)
        }
    }

    override fun getCount(): Int {
        synchronized(appList) {
            return appList.size
        }
    }

    override fun getItem(arg0: Int): Any {
        synchronized(appList) {
            return appList[arg0]
        }
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    @SuppressLint("InflateParams", "ViewHolder")
    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {
        val appInfo = getItem(position) as AppInfo
        val view = LayoutInflater.from(ctx).inflate(R.layout.item_applist, null)
        val viewHolder = ViewHolder()
        viewHolder.icon = view.findViewById(R.id.app_icon)
        viewHolder.appName = view.findViewById(R.id.app_name)
        viewHolder.desc1 = view.findViewById(R.id.app_desc1)
        viewHolder.desc2 = view.findViewById(R.id.app_desc2)
        view.tag = viewHolder
        try {
            viewHolder.appName?.text = appInfo.appName
            viewHolder.desc1?.text = appInfo.packagePath
            var icon = appInfo.icon
            if (icon == null) {
                icon = appInfo.applicationInfo.loadIcon(pm)
            }
            viewHolder.icon?.setImageDrawable(icon)
        } catch (t: Throwable) {
            t.printStackTrace()
        }
        return view
    }

    private enum class AppListOrder {
        BY_NAME, BY_INSTALL_TIME
    }

    internal class ViewHolder {
        var icon: ImageView? = null
        var desc2: TextView? = null
        var desc1: TextView? = null
        var appName: TextView? = null
    }
}